code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Offer.kt

```diff
@@ -0,0 +1,279 @@
+
+            val data = mutableMapOf<String, Any>()
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_ID] = id
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_TYPE] = type.toString()
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CONTENT] = content
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_LANGUAGE] = language
+            data[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CHARACTERISTICS] = characteristics
+
+            offerMap[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA] = data
+
+            return offerMap
+        }
+
+        private fun getContentFromOfferData(offerData: Map<String, Any?>): String {
+            val offerContent = offerData[OptimizeConstants.JsonKeys.PAYLOAD_ITEM_DATA_CONTENT]
+            return when (offerContent) {
+                is String -> offerContent
+                is Map<*, *> -> JSONObject(offerContent).toString()
+                else -> ""
+            }
+        }
+
+        fun List<Offer>?.displayed() {
+            OptimizeUtils.trackWithData(this.generateDisplayInteractionXdm())
+        }
+
+        fun List<Offer>?.generateDisplayInteractionXdm(): Map<String, Any>? {
+            return OptimizeUtils.generateInteractionXdm(
+                OptimizeConstants.JsonValues.EE_EVENT_TYPE_PROPOSITION_DISPLAY,
+                this?.toPropositionsList()
+            )
+        }
+
+        private fun List<Offer>?.toPropositionsList(): List<OptimizeProposition>? {
+            val propositions = this
+                ?.mapNotNull { it.propositionReference?.get() }
+                .takeIf { !it.isNullOrEmpty() }
+            return propositions?.filter { proposition ->
+                proposition.offers.any { offerInProp ->
+                    this?.any { it.id == offerInProp.id } == true
+                }
+            }
+        }
+    }
+
+    var id: String = ""
+        private set
+    var etag: String = ""
+        private set
+    var score: Double = 0.0
+        private set
+    var schema: String = ""
+        private set
+    var meta: Map<String, Any> = HashMap()
+        private set
+    var type: OfferType = OfferType.UNKNOWN
+        private set
+    var language: List<String> = ArrayList()
+        private set
+    var content: String = ""
+        private set
+    var characteristics: Map<String, String> = HashMap()
+        private set
+    var propositionReference: SoftReference<OptimizeProposition>? = null
+
+    class Builder(private val id: String?, private val type: OfferType?, private val content: String?) {
+        private val offer = Offer()
+        private var didBuild = false
+
+        init {
+            offer.id = id ?: ""
+            offer.type = type ?: OfferType.UNKNOWN
+            offer.content = content ?: ""
+        }
+
+        fun setEtag(etag: String?) = apply {
+            throwIfAlreadyBuilt()
+            offer.etag = etag ?: ""
```

Comment:
Etag can we nullable. Why are we setting the value to empty is if its null?

Reply:
Also what happened to all the in line docs? Why did we remove it? All these ate public methods